---
title: "Clustering analysis of differentially expressed genes in the BRAUN Pair: \n DE in TR-FP"
author: "Eric Durandau"
# date: "07/10/2020"
output: 
  flexdashboard::flex_dashboard:
  # storyboard: true
  social: menu
source: embed
# widescreen: true
# fig_width: 7
# fig_height: 6
# fig_caption: true
df_print: kable
# css: https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css
---
  <style type="text/css">
  .centered{
    float:right;
    text-align:center;
    width:100%;
  }

  </style>
  
```{r BRAUNsetup global, include=FALSE}
OutputFolder="TR_V_FP"
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(echo = FALSE,
                      fig.path = paste(OutputFolder,"Figures/", sep = "/"),
                      dev = "svg")

```

```{r BRAUNsetup Anal, echo=F, warning=FALSE, message=FALSE}
library(data.table)
library(knitr)
library(kableExtra)
library(ggsignif)
library(ggpubr)
library(grid)
library(gridExtra)
library(RColorBrewer)

RootFolder=dirname(rstudioapi::getSourceEditorContext()$path)
setwd(RootFolder)
# source various R script to import functions
# 1. Source the file containing the home made functions
source(file.path("..", "R", "RTools.R"))
source(file.path("..", "R","20200610-DGE_library_functions.R"))
source(file.path("..", "R","GO.R"))

# Check Ca GO db
Data_dir=file.path("..","Data")

if( ! require( "org.Calbicans.eg.db" , character.only = TRUE ) ){
  BuildCaGOdb_CP(OuputDir = Data_dir)
  install.packages(file.path(Data_dir,"org.Calbicans.eg.db"),repos = NULL)
}
```

```{r BRAUNsetup exp, include=FALSE}
Pair="BRAUN"
# Path and file names
# root folder
# RootFolder="/media/edurandau/DATA/Eric_Durandau/Sanglard/Labbook/Workbench/Bioinformatics/GREEN-BRAUN-data_exploration/Tests"
# Output subfolder name
# OutputFolder="Hs_3-6h_V_0h"
# Outputfolder path
# OutputFolderPath=file.path(RootFolder,OutputFolder)

# Raw data table
Path2Data=file.path(Data_dir,"20210109-Ca_GREEN_BRAUN_RawRead_Count.tab")
# Table containing the list of sample ID and coresponding real names
Path2SubTbl=file.path(Data_dir,"20200723-LIMS_Sample_Submission.tbl")
# Read count table with corrected sample names
RC_correctedSampleNames=paste0(Pair,"CountTable_SampleNames.tbl",collapse = "_")

# Variable related to Normalisation, Differential expression analysis, clustering and the GO part
# NORMALISATION
# RegEx pattern enabling to extract the MetaData from samples real names
RegexPattern="(?<Isolate>\\w{3}\\d{4})_(?<replicate>\\d{1})_(?<condition>\\w{2})_(?<time>\\d{1})"
# RegEx pattern enabling to remove some sample from the analysis
# ExcludeSampleRegex="(?<Isolate>\\w{3}(3621|3678))_(?<replicate>\\d+)"
ExcludeSampleRegex="(?<Isolate>\\w{3}(3678|3621))_(?<replicate>\\d+)"
# Metadata column to conserve in the "Genes" table. Renames the column to make it compatible with HomeMade functions
GenesMetaData=c("Gene", "ORF19_ID","ASSEMBLY22_ID","Length","EntrezID","Note")
Colnames4Genes=GenesMetaData
# Additional factor for the normalisation and DE
AdditionalFact=c("Isolate","time", "condition")
# DEG
# Variable for the contrast map to define the comparaison to perform
DE_Contrast_Cond=AdditionalFact
Ref=c("CEC3605", 0,"FP")
# Sample="Conditions"
# Select the main comparaison variable
MainComparaison="condition"
SecondaryComparaison=AdditionalFact[1:2]
# define the log2FC cutoff and the pvalue cutoff
LFC_Cutoff=0.5
pVal_Cutoff=0.05
# Clustering
# Number of cluster
ClustNumb=4
# distance method
DistMethod="manhattan"
# agglomeration method
# AggloMethod="ward.D"
AggloMethod="centroid"
# Decide Cutoff. for example if 3 , gene is selected if differentially expressed in 1/3 of conditions
Factor_Numb_Conditions_gene_diff_expressed=3

#  GO enrichment
# library(org.Calbicans.eg.db)
OrgDb=org.Calbicans.eg.db::org.Calbicans.eg.db
keytype="GID"
GO_pValue=0.2
GO_qValue=GO_pValue
MinGO=10
MaxGO=50
# STRING
STRING_DB=string_db_Ca
StringConnection=50

# for Candida

# Setup procedure
#  create the directory if require and set the folder as working directory
# dir.create(OutputFolderPath,showWarnings = F)
```


*Candida albicans* {data-navmenu="Normalisation" .storyboard}
=====================================

### Normalisation 

```{r BRAUN change sample IDs by sample real names, echo=F, warning=FALSE, message=FALSE}
# 5. Import the Sample submission table
# setwd(OutputFolderPath)
SampleID=read.delim(Path2SubTbl,sep = "\t",row.names = 2, header = F)
# 6. Correct readcount name
RC_Raw=read.delim(Path2Data,sep = "\t",header = T,skip = 1)
RC_ColVar=colnames(RC_Raw)
GetCoordStr=function(ToFind){
  Coord=which(str_detect(RC_ColVar, paste0(ToFind,".bam")))
}
Coor=rapply(as.list(rownames(SampleID)), GetCoordStr)
colnames(RC_Raw)[Coor]=SampleID$V1
# 7.save the corrected readcount table
write.table(RC_Raw, file= file.path(Data_dir,RC_correctedSampleNames), row.names =F, sep = "\t")

```

```{r BRAUNImport and normalise, echo=F, warning=FALSE, message=FALSE, fig.show='hide',results='hide'}
# 8. Import the data
RawData=LoadRawReadCounts(file.path(Data_dir,RC_correctedSampleNames), MetaDataFromSampleName=RegexPattern, ExcludeSamples=ExcludeSampleRegex, skip = 0)
# # before starting the normalisation we reduce the Gene table in order to conserve key informations
Genes=RawData$Genes[,GenesMetaData]
# modify variable names to make the gene table compatible with next functions
colnames(Genes)= Colnames4Genes

NormData=ReadCountNormalisation(RawData$ReadCounts, RawData$MetaData, RawData$MetaData$Conditions, Genes, AdditionalFactor=AdditionalFact, Plots = T)
# NormData$GGplot$Batch_effect_density

# grid.arrange(arrangeGrob(grobs=NormData$GGplot$Batch_effect_heatmap ,ncol=2))

```

  
#### **Change Samples IDs by Sample names and save the table**
  
Using the table furnished by Kontxi Martinez, Samples IDs are replaced by the experiment name in the *read courn* table. The experiment name follow the current pattern:
  
- **Isolate**_**Replicate**_**Condition**_**Harvesting time**
  
#### **Import the corrected table**
  
The corrected *read count* table is imported. Only value of the `r Pair` pair are kept. The homemade function *LoadRawReadCounts()* is used for this.

#### **Normalisation**
  
Read count are normalized using the *ReadCountNormalisation()* homemade function
  
- [ ] list of packages used
- [ ] better description of the global proc√©dure


### Low reads filtering

```{r BRAUNlow expressed genes,echo=F, warning=FALSE, message=FALSE,  fig.height=8, fig.width=13}
NormData$GGplot$Low_Expression_Filtering
```

***
The first step consist in removing features, here genes, with no or low reads in all samples. The sequencing detection limit is estimated at 10. This means that genes with less than 10 reads in all samples are considered as not expressed (or expressed under the detection limit) and are thus removed from the analyis. Here, `r sum(rowSums(NormData$Expressed_Genes)==0)` genes have been rejected. The filtering is based on the the [filterByExpr](https://rdrr.io/bioc/edgeR/man/filterByExpr.html) from the [edgeR](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf) packgage. The figure shows the density of gene read counts. Note the vertical line indicating the threshold used to sort out none expressed genes.

### Sequencing depth

```{r BRAUNsequencing depth,echo=F, warning=FALSE, message=FALSE,  fig.height=8, fig.width=13}
NormData$GGplot$Library_size
```

***
Barplot shows the whole number of read counts for each sample. Reads which are not associated to an annotated gene, are not considered in the calculation. 

A filtering is also applied to identify and remove genes with less than 10 reads (see the [Low read filtering](#Low_reads_filtering) tab). Reads associated to these filtered out genes are rejected.


### TMM based data correction

```{r BRAUNTMM correction,echo=F, warning=FALSE, message=FALSE, fig.height=8, fig.width=13}
NormData$GGplot$TMM_Correction
```

***
This step consist in removing the compositions bias using the TMM (Trimed mean  of M-value) [Robinson and Oshlack 2010](8https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25). In short, samples with different library size cannot be compared with each other. The process consist in normalise the read count such like each sample have artificially the same total number of reads.

The [calcNormFactors](https://www.rdocumentation.org/packages/edgeR/versions/3.14.0/topics/calcNormFactors) from the [edgeR](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf) package has been used to perform this process.

boxplot represent the distribution of genes **log2(cpm)**. **cpm** are *count per milion* of read.


### Voom transform of read counts

```{r BRAUNvoom transform,echo=F, warning=FALSE, message=FALSE, fig.height=8, fig.width=13 }
NormData$GGplot$Voom_transform

```

***

**Voom transform of counts : removing heteroscedascity from count data**
  
The basic principle is to remove a bias related to the relationship between gene expression variance and gene expression mean. The more condition/replicate we have the better it is. *voom* aim at prepare the data for a further linear fit see chapter 6.2 of [RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR](https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html). This operation is performed using the [voom](https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/voom) function from the [limma](https://www.bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf) package. 

For a given data set with multiple samples, it is expected the average read count associated to a genes should not depend on its variance. More simply, the variability among samples should be similar for every genes and not depends on their average expression level. For RNAseq data, this relationship between variability and expression is unfortunaltely true (figure right). Lower will be global number of reads associated to a gene, higher will be the variability among samples. This may have an impact on p-value when comparing the expression of gene between two samples. 

Voom transform attempt to correct for this bias. From the comparison of the standard deviation against the average read count of each gene, a model is built (right panel, red line) and used to correct read counts value (left panel).
  
### Batch effect correction: PCA

```{r BRAUNbatch PCA plot,echo=F, warning=FALSE, message=FALSE, fig.height=4, fig.width=10 }
ggplotly(NormData$GGplot$Batch_effect_PCA,width =1400 , height = 700)
```

***
This step might not be essential since all samples are from the same batch. This can be interesting if sequencing files come from different runs or lanes. We will use the [SVA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3307112/) package to infer the batch effects. Very quickly sva estimate the number of covariate (surrogate variable) and then adjust value to remove the noise associated to such covariate. Then adjust for the batch effects using Surrogate variables. Here the strategy consist in let the algorithm estimate the potential batch variables rather than feeding it with known source of batch effect.

Different visualisation can be used to observe the correction of the batch effect. The first is the PCA. PCA plot is a scatter where each dot is a sample. Ideally, all replicates from a given condition should appears close from each other, indicating a good reproducibility. Our strategy based on the estimation of variables inducing noise may reduced the distance between replicates on the PCA plot.

### Batch effect correction: Density

```{r BRAUNbatch density plot,echo=F, warning=FALSE, message=FALSE, fig.height=4, fig.width=10 }
ggplotly(NormData$GGplot$Batch_effect_density,width =1400 , height = 700)
```

*** 
The second visualisation is the density plot. The is basically the density of gene count relative to the read count range (or log2(cpm)). The plot confirm the quality of the whole normalisation since all samples show similar density.

### Batch effect correction: Correlation matrix

```{r BRAUNbatch effect heatmap, echo=F, warning=FALSE, message=FALSE, fig.height=8, fig.width=15 }
grid.arrange(arrangeGrob(grobs=NormData$GGplot$Batch_effect_heatmap ,ncol=2))
```

***
The last representation is the correlation matrix. A correlation coefficient is calculated between each samples combination. The resulting matrix of correlation coeficients is then visualised using a heatmap. A layer of Hierarchical clustering analysis has been performed on this matrix. This enables to sort rows and columns depending on the similarity of their correlation coefficients. Ideally the Heatmap should show samples from similar conditions close from each other. Note the improvement of the clustering after the batch effect correction. Note also that samples from similar experimental conditions cluster together, especially depending on the harvesting time. 

<!-- # Analysis setup and Normalisation -->

<!-- ```{r BRAUNsetup, include=FALSE} -->
<!-- OutputFolder="TR_V_FP" -->
<!-- knitr::opts_chunk$set(echo = FALSE) -->
<!-- knitr::opts_chunk$set(echo = FALSE, -->
<!--                       fig.path = paste(OutputFolder,"Figures/", sep = "/"), -->
<!--                       dev = "svg") -->

<!-- ``` -->


<!-- - Import the Read Count table, keep the count related to the `r Pair` pair. Use the *LoadRawReadCounts()* -->



# Differentially expressed genes
1. Prepare a table containing information about how to perfom compararaison between samples
```{r BRAUNDE gene : comparaison table,echo=F, warning=F, message=FALSE, results='hide'}
# first create a table mapping the which define condition for which DE analysis has to be performed
  # recover the metadata from the "RawData"object
Meta=RawData$MetaData
  # remove replicate infos define the condition to look at and the corresponding reference Value
Meta2=Meta[!duplicated(Meta$Conditions),]
Cond=Meta2[,c("Conditions", DE_Contrast_Cond)]

  # Build the comparaison map using the "BuildDEComparaison" fonction
Map=BuildDEComparaison(Cond,"Conditions", Ref)
PMap=Map
PMap$Name=str_replace(PMap$Name, "_-_", " V ")
PMap
```
2. Perform the differencial expression (DE) analysis using the *Genes_Diff_Expressed*
  - Search for differentially expressed genes for all comparaisons listed in the above table
  - p-value cutoff at `r pVal_Cutoff` and log2(FC) cutoff at `r LFC_Cutoff` (>`r 2^LFC_Cutoff` fold change)
  - Find the result in the [`r OutputFolder`/Global_DEG](./`r OutputFolder`/Global_DEG) subfolder

```{r BRAUN DE gene , echo=F, warning=F, message=FALSE, results='hide'}
# Perform Differencial expression analysis
DETest=Genes_Diff_Expressed(NormData$Norm_Data,NormData$Norm_Data$targets$group_Cond,Group4InteractivePlot = data.frame(NormData$Norm_Data$targets[SecondaryComparaison]), AdditionalFactor=SecondaryComparaison, RefCond = Map$Ref[Map$Test_Cond==MainComparaison],CompareCond = Map$Test[Map$Test_Cond==MainComparaison],Method = "Paired",lfc_cuttoff = LFC_Cutoff, pvalue_cuttoff = pVal_Cutoff, Plots=file.path(OutputFolderPath, "Global_DEG"))

DETest=Genes_Diff_Expressed(NormData$Norm_Data,NormData$Norm_Data$targets$group_Cond,Group4InteractivePlot = data.frame(NormData$Norm_Data$targets[SecondaryComparaison]), AdditionalFactor=SecondaryComparaison, RefCond = Map$Ref[Map$Test_Cond==MainComparaison],CompareCond = Map$Test[Map$Test_Cond==MainComparaison],Method = "Paired",lfc_cuttoff = LFC_Cutoff, pvalue_cuttoff = pVal_Cutoff)

Map
```

**Summary statistics :**

```{r BRAUNprint Summary, echo=F, warning=F, message=FALSE }
SumupBay=data.table::as.data.table(DETest$eBayes_Summary)
test=as.data.frame(dcast(SumupBay, V1~V2))
colnames(test)=str_replace(colnames(test), "-", " V ")
test%>%kable("html",caption = paste0("p-value<",as.character(pVal_Cutoff),", log2(FC)>",LFC_Cutoff)) %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, full_width = F)

```


# Selected hierachical clustering method using logFC values

<div class='left' style='float:left;width:40%;'>

- Reduce dimension Using [UMAP](https://umap-learn.readthedocs.io/en/latest/)
- Distance type : `r DistMethod`
- Agglomeration : `r AggloMethod`
- #Clusters : `r ClustNumb`

**Find the result in the [`r OutputFolder`/HC_Clustering](./`r OutputFolder`/HC_Clustering) subfolder**

```{r BRAUNDE Genes statistics, echo=F, warning=F, message=FALSE }
Decide=decideTests(DETest$eBayes_Stat,adjust.method = "BH",p.value = pVal_Cutoff, lfc = LFC_Cutoff)

Bol=!(rowSums(abs(Decide)))<ncol(Decide)/Factor_Numb_Conditions_gene_diff_expressed
LFC=as.data.frame(DETest$eBayes_Stat$coefficients[Bol,])
pVal=as.data.frame(DETest$eBayes_Stat$p.value[Bol,])
# head(LFC)y¬ß
PCA=M3C::umap(t(LFC))
PCA$layout=PCA$data

# ClustNumb=6
Dist=dist(PCA$layout,method = DistMethod)

Clust=cutree(hclust(Dist, method = AggloMethod),k=ClustNumb)

PCA_Data=as.data.frame(cbind(PCA$layout, Clust))
PCA_Data$Clust=as.factor(PCA_Data$Clust)
# P=ggplot(PCA_Data, aes(x=PC1, y = PC2, color=Clust)) +
P=ggplot(PCA_Data, aes(x=X1, y = X2, color=Clust)) +
      geom_point()+
      ggtitle(paste(DistMethod,"distance \nwith" , AggloMethod, "agglomeration")) +
  mynamesthemeSmall +
  scale_color_brewer(palette="Paired")
# ggsave("/home/edurandau/Desktop/temp/00000f.png", P,width = 7)
  # scale_color_brewer(palette="Dark2")

# add cluster information to logFod change table
LFCClust=cbind(LFC,Clust)


# convert to data table
# DTMeta=merge(DETest$MetaData,Meta2[,c("Sample", AdditionalFact)], by.x = )
DTMeta=as.data.table(DETest$MetaData, keep.rownames = T)
DTLFC=as.data.table(LFCClust,keep.rownames = T)
# reshape the LFC table
DTLFC=melt(DTLFC,id.vars = c("rn", "Clust"))

# fuse table adding condition and time to the LFC matrix
DTfuseMetaNLFC=DTMeta[DTLFC,on =.(rn=variable)]
# rename columns
setnames(DTfuseMetaNLFC, c("Comparaison",colnames(DETest$MetaData), "Gene", "Cluster","Log.FC"))

SumStatDE=DTfuseMetaNLFC[,.(Number.Gene=length(unique(Gene)),
                               Cluster.Mean.LogFC = mean(Log.FC) ,
                               SD = sd(Log.FC),
                            Gene=paste(unique(Gene),collapse = ","))
                         , by=Cluster]
# Prepare a summary table
IntermedStat=DTfuseMetaNLFC[, .(Mean.LogFC = mean(Log.FC) ,
                                SD = sd(Log.FC),
                           Cluster=mean(Cluster))
                         , by=Gene]
SumUPFromInter=IntermedStat[,.(Number.Gene=.N,
                              Mean.LogFC = mean(Mean.LogFC) ,
                               SEM = sd(Mean.LogFC),
                            Gene=paste(Gene, collapse = ","))
                         , by=Cluster]

SumStatDE[,1:(ncol(SumStatDE)-1)]%>%kable("html",caption = paste("Clustering of Gene set with log fold change >", LFC_Cutoff)) %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, full_width = F)


dir.create(file.path(OutputFolderPath, "HC_Clustering"),showWarnings = F)
write.table(SumStatDE, file = file.path(OutputFolderPath, "HC_Clustering", "Gene_list.tbl"), sep = "\t", row.names=FALSE)

# add moderate t statisitc pvalue to the LFC table
LFCClustNpVal=cbind(LFCClust,DETest$eBayes_Stat$F.p.value[Bol])
colnames(LFCClustNpVal)[ncol(LFCClustNpVal)]="Mod_Fstat_pVal"

# export the cluster table
# Add gene information to the LFC table
FullTble=merge(Genes,round(LFCClustNpVal,2),by="row.names")


for (i in 1:ClustNumb){
  write.table(FullTble[FullTble$Clust==i,],
              file = file.path(OutputFolderPath, "HC_Clustering", paste0("Gene_list_", as.character(i),".tbl")),
              sep = "\t",
              row.names=FALSE)
}
write.table(FullTble,
              file = file.path(OutputFolderPath, "HC_Clustering", paste0("DE_Genes_N_Clust.tbl")),
              sep = "\t",
              row.names=FALSE)
print(summary(aov(Log.FC~Cluster,data = DTfuseMetaNLFC)))
# print(summary(aov(Mean.LogFC~Cluster,data = IntermedStat)))


```

</div>
<div class='right' style='float:right;width:60%;'>
<div class='left' style='float:left;width:40%;'>

```{r BRAUNselected HC method, echo=F, warning=F, message=FALSE, results='hide',fig.height=7, fig.width=3.5}

Generate_list_Stat_Compare=function(x){
  Cond=as.character(unique(DTfuseMetaNLFC[[x]]))

  ListComp=list()
  Ind=1
  for (i in 2:length(Cond)){
    ListComp[[Ind]]=as.character(c(Cond[1], Cond[i]))
    Ind=Ind+1
  }
  D=data.table::copy(DTfuseMetaNLFC[bol,])
  setnames(D, old = x, new = "Var")

  Formula=as.formula(paste("~", paste(Condlabel[Condlabel!=x], collapse = "+")))
  P=ggplot(D, aes(x=Var,y=Log.FC,fill=Var)) +
  geom_violin() +
  stat_compare_means( comparisons =ListComp , method = "t.test", label = "p.signif") +
  mynamesthemeSmall+ xlab(x)+facet_wrap(Formula)+
   theme(legend.position="none")

  return(P)
}
Condlabel=colnames(DETest$MetaData)
bol=rep(T,nrow(DTfuseMetaNLFC))
Violin_plots=lapply(Condlabel, Generate_list_Stat_Compare)
Violin_plots[[length(Violin_plots)+1]]=P
Test=grid.arrange(grobs=Violin_plots, ncol=1)


```
</div>
<div class='right' style='float:right;width:60%;'>

```{r BRAUNselected HC method PCA, echo=F, warning=F, message=FALSE, results='hide', fig.height=7, fig.width=5}
Index=sort(as.numeric(PCA_Data$Clust), decreasing = T, index.return=T)
HC_Cluster=as.data.frame(Clust[Index$ix])
colnames(HC_Cluster)="HC_Cluster"

Annot=droplevels(DETest$MetaData)

Colorlist=brewer.pal(12, "Paired")
AnnotCol=sapply(colnames(Annot), function(x){
  RandNum=sample.int(12, length(levels(Annot[[x]])))
  Color=Colorlist[RandNum]
  names(Color)=levels(Annot[[x]])
  Color
}, simplify = FALSE,USE.NAMES = TRUE)
AnnotCol[["HC_Cluster"]]=Colorlist[1:length(levels(PCA_Data$Clust))]
names(AnnotCol[["HC_Cluster"]])=levels(PCA_Data$Clust)
# P1= pheatmap(as.matrix(LFC[Index$ix,]),cluster_cols = T, cluster_rows = F, main = "Canberra distance \nwith average agglomeration", width = 20, height = 60, annotation_row =HC_Cluster ,annotation_names_row = F ,annotation_col = DETest$MetaData, fontsize_row =2)
P1=pheatmap(as.matrix(LFC[Index$ix,]),cluster_cols = T, cluster_rows = F, main = paste(DistMethod,"\nwith",AggloMethod,"agglomeration") , width = 20, height = 60, annotation_row =HC_Cluster ,annotation_names_row = F ,annotation_col = Annot, fontsize_row =2, annotation_colors = AnnotCol)


# par(mfrow=c(1,1))
# P1
# Test
```
</div>
</div>

# HC Gene Clusters Features : Clusters 1 - 4
<div class='centered'>
```{r BRAUNclustering HC, echo=F, warning=F, message=FALSE, results='hide',fig.height= 7, fig.width= 12}
PlotList= list()

PCA_Data=as.data.frame(cbind(as.data.frame(PCA$layout), Clust))

for (i in unique(Clust)){
  SClust=Clust

  SClust[SClust!=i]=0
  PCA_Data[,"SClust"]=as.factor(SClust)

  P1=ggplot(PCA_Data, aes(x=X1, y = X2)) +
      geom_point(aes(colour=SClust), show.legend = FALSE)+
    scale_color_manual(values=c("grey", Colorlist[i]))+
    mynamesthemeSmall

  bol=DTfuseMetaNLFC$Cluster==i
  Violin_plots=lapply(Condlabel, Generate_list_Stat_Compare)
  Violin_plots=c(list(P1),Violin_plots)

  PlotList[[paste0("plot",as.character(i))]]=arrangeGrob(grobs=Violin_plots, nrow=length(Violin_plots),)
}

grid.arrange(grobs=PlotList[1:4], ncol=4)
# P

```
</div>

<!-- # HC Gene Clusters Features : Clusters 5 - 8 -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNclustering HC 2, echo=F, warning=F, message=FALSE, results='hide',fig.height= 7, fig.width= 12} -->
<!-- grid.arrange(grobs=PlotList[5], ncol=4) -->

<!-- ``` -->
<!-- </div> -->

<!-- # HC Gene Clusters Features : Clusters 9 - 12 -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNclustering HC 3, echo=F, warning=F, message=FALSE, results='hide',fig.height= 7, fig.width= 12} -->
<!-- grid.arrange(grobs=PlotList[9:12], ncol=4) -->
<!-- ``` -->
<!-- </div> -->

# Summary table of differentially expressed genes

```{r BRAUNSummary table DEG 3hrs, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" }
ST=merge(Genes[, c("Gene", "Note")],round(LFCClustNpVal,2),by="row.names")
ST=ST[, 2:ncol(ST)]
ST$Clust=as.integer(ST$Clust)
DT::datatable(ST, filter = "top",extensions = 'Buttons', options = list(scroll = TRUE,dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
  headerCallback = DT::JS(
    "function(thead) {",
    "  $(thead).css('font-size', '0.7em');",
    "}"
  )
), rownames = F)%>%
  DT::formatStyle(columns = colnames(ST), fontSize = '75%')
```
# GO analysis using "Cluster Profiler"

<!-- ```{r BRAUNGO Annotation , echo=F, message=FALSE, warning=FALSE, results='hide' } -->
<!-- Genes$HC_Clusters=0 -->
<!-- Genes$MeanLFC=0 -->
<!-- MeanLFC=rowMeans(LFC) -->
<!-- coord2=match(names(MeanLFC),rownames(Genes)) -->
<!-- Genes$MeanLFC[coord2]=MeanLFC -->

<!-- coord=match(rownames(HC_Cluster), rownames(Genes)) -->
<!-- Genes$HC_Clusters[coord]=HC_Cluster$HC_Cluster -->

<!-- Genes=Genes -->

<!-- PlotList=lapply(unique(Genes$HC_Clusters), function(x){ -->
<!--   # library(org.Hs.eg.db) -->
<!--   # OrgDb=org.Hs.eg.db::org.Hs.eg.db -->
<!--   ClusterNb=x -->
<!--   if (x==0){ -->
<!--     GeneList=Genes$EntrezID[Genes$HC_Clusters!=0] -->
<!--     GeneFC=Genes$MeanLFC[Genes$HC_Clusters!=0] -->
<!--     ClusterNb="All_DE" -->
<!--   }else{ -->
<!--     GeneList=Genes$EntrezID[Genes$HC_Clusters==ClusterNb] -->
<!--     GeneFC=Genes$MeanLFC[Genes$HC_Clusters==ClusterNb] -->
<!--   } -->
<!--   names(GeneFC)=GeneList -->

<!--   Plot=GOE(GeneFC,as.character(Genes$EntrezID), OrgDb,keyType =keytype , pvalueCutoffE = GO_pValue , qvalueCutoffE = GO_qValue,MinGOCount=MinGO, MaxGOCount=MaxGO) -->
<!--   dir.create(file.path(OutputFolderPath, "GO"), showWarnings = F) -->
<!--   # dir.create(file.path(OutputFolderPath, "GO",Comparison)) -->

<!--   write.table(Plot[["Summary"]], -->
<!--               file = file.path(OutputFolderPath, "GO", paste0("Cluster",as.character(x),"_GO2Gene.tbl")), -->
<!--               sep = "\t", -->
<!--               row.names=FALSE) -->
<!--   return(Plot) -->
<!-- }) -->
<!-- names(PlotList)=unique(Genes$HC_Clusters) -->

<!-- Genes2=Genes -->
<!-- Genes2=Genes2[!is.na(Genes2$EntrezID),] -->

<!-- StringMap <- STRING_DB$map( Genes2, "EntrezID", removeUnmappedRows = TRUE ) -->

<!-- ``` -->
<!-- 1. perform GO enrichment using (ClusterProfiler)[http://yulab-smu.top/clusterProfiler-book/] -->
<!-- 2. Simplify the GO term list by removing redundant elements -->



<!-- # Cluster 1 : GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C1 GO list, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=1 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 1 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC1 gene Network, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->


<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 1 : Enrich map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC1 gene Enrich D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->


<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 1: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC1 interaction map, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 1 : GO to Gene Association -->

<!-- ```{r BRAUNC1 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- } -->
<!-- ``` -->


<!-- # Cluster 2 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C2 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=2 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 2 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC2 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 2: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC2 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 2: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC2 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 2 : GO to Gene Association -->

<!-- ```{r BRAUNC2 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- } -->
<!-- ``` -->



<!-- # Cluster 3 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C3 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=3 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 3 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC3 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 3: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC3 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 3: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC3 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 3 : GO to Gene Association -->

<!-- ```{r BRAUNC3 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- } -->
<!-- ``` -->


<!-- # Cluster 4 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C4 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=4 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 4 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC4 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->


<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 4: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC4 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 4: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC4 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 4 : GO to Gene Association -->

<!-- ```{r BRAUNC4 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- } -->
<!-- ``` -->


<!-- # Cluster 5 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C5 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=5 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 5 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC5 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 5: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC5 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 5: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC5 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 5 : GO to Gene Association -->

<!-- ```{r BRAUNC5 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- ``` -->

<!-- # Cluster 6 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C6 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=6 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 6 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC6 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 6: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC6 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 6: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC6 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 6 : GO to Gene Association -->

<!-- ```{r BRAUNC6 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- ``` -->

<!-- # Cluster 7 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C7 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=7 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 7 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC7 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 7: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC7 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 7: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC7 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 7 : GO to Gene Association -->

<!-- ```{r BRAUNC7 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- ``` -->

<!-- # Cluster 8 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C8 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=8 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 8 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC8 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 8: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC8 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 8: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC8 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 8 : GO to Gene Association -->

<!-- ```{r BRAUNC8 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- ``` -->

<!-- # Cluster 9 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C9 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=9 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 9 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC9 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 9: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC9 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 9: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC9 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 9 : GO to Gene Association -->

<!-- ```{r BRAUNC9 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- ``` -->

<!-- # Cluster 10 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C10 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=10 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 10 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC10 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 10: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC10 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 10: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC10 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 10 : GO to Gene Association -->

<!-- ```{r BRAUNC10 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- ``` -->

<!-- # Cluster 11 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C11 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=11 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 11 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC11 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->

<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 11: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC11 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 11: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC11 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 11 : GO to Gene Association -->

<!-- ```{r BRAUNC11 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- ``` -->

<!-- # Cluster 12 GO List -->
<!-- <div class='centered'> -->
<!-- ``` {r C12 GO list D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- C=12 -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=1 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 12 : Gene Network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC12 gene Network D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->


<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=3 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->
<!-- ``` -->
<!-- </div> -->

<!-- # Cluster 12: Enriched map -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC12 enriched map D, fig.height= 8, fig.width= 18, echo=F, message=FALSE, warning=FALSE, results='hide'} -->

<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->
<!-- PlotNb=4 -->
<!-- # TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!-- if (length(PlotList[[as.character(C)]])<2){ -->
<!--   print("No GO emerge from that list of gene") -->
<!-- }else{ -->
<!--   TooPlot=sapply(PlotList[[as.character(C)]],"[[",PlotNb,simplify = F) -->
<!--  ggarrange(plotlist=TooPlot[1:length(TooPlot)-1], -->
<!--                 ncol = 3, -->
<!--                 common.legend = F -->
<!--  ) -->
<!-- } -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 12: Protein interaction network -->
<!-- <div class='centered'> -->
<!-- ```{r BRAUNC12 interaction map D, fig.height= 10, fig.width= 10, echo=F, message=FALSE, warning=FALSE, results='hide'} -->
<!-- PlotSTRINGdbNetwork(StringMap,string_db = STRING_DB,Cluster = C, MaxConnection = StringConnection) -->

<!-- ``` -->
<!-- </div> -->


<!-- # Cluster 12 : GO to Gene Association -->

<!-- ```{r BRAUNC12 Summary GO D, message=FALSE, warning=FALSE,out.width = "30%", out.height = "10px" } -->
<!-- #CN=which(str_detect(names(PlotList),paste0("_",as.character(C)))) -->


<!-- DT::datatable(PlotList[[as.character(C)]][["Summary"]], filter = "top",extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), -->
<!--   headerCallback = DT::JS( -->
<!--     "function(thead) {", -->
<!--     "  $(thead).css('font-size', '0.7em');", -->
<!--     "}" -->
<!--   ) -->
<!-- ), rownames = F)%>% -->
<!--   DT::formatStyle(columns = colnames(PlotList[[as.character(C)]][["Summary"]]), fontSize = '75%') -->
<!-- ``` -->
