<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Transcriptomics analysis of pairs of Candida albicans clinical isolates</title>

<script src="site_libs/header-attrs-2.9/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Transcriptomics analysis of pairs of Candida albicans clinical isolates</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    BRAUN
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Candida albicans</li>
    <li>
      <a href="BRAUN_ca_TRVFP.html">TR VS FP</a>
    </li>
    <li>
      <a href="BRAUN_ca_time.html">3-6hrs. VS 0hr.</a>
    </li>
    <li>
      <a href="BRAUN_ca_3617V3605.html">CEC3617 V CEC3605</a>
    </li>
    <li class="dropdown-header">Homo sapiens</li>
    <li>
      <a href="BRAUN_Hs_time.html">3-6hrs. VS 0hr.</a>
    </li>
    <li>
      <a href="BRAUN_Hs_3617V3605.html">CEC3617 V CEC3605</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    GREEN
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Candida albicans</li>
    <li>
      <a href="GREEN_ca_TRVFP.html">TR VS FP</a>
    </li>
    <li>
      <a href="GREEN_ca_time.html">3-6hrs. VS 0hr.</a>
    </li>
    <li>
      <a href="GREEN_ca_3678V3621.html">CEC3678 V CEC3621</a>
    </li>
    <li class="dropdown-header">Homo sapiens</li>
    <li>
      <a href="GREEN_Hs_time.html">3-6hrs. VS 0hr.</a>
    </li>
    <li>
      <a href="GREEN_Hs_3678V3621.html">CEC3678 V CEC3621</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    BLUE
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Candida albicans</li>
    <li>
      <a href="BLUE_ca_TRVFP.html">TR VS FP</a>
    </li>
    <li>
      <a href="BLUE_ca_time.html">6-24hr VS 3hr</a>
    </li>
    <li>
      <a href="BLUE_ca_3609V3672.html">CEC3609 V CEC3672</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Transcriptomics analysis of pairs of <em>Candida albicans</em> clinical isolates</h1>

</div>


<pre class="r"><code>InstallPackages=function(x){
  if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE)) install.packages(&quot;BiocManager&quot;)
  for( i in x ){
    #  require returns TRUE invisibly if it was able to load package
    if( ! require( i , character.only = TRUE ) ){
      # check the package is a bioconductor package and install it if required
      if (length(BiocManager::available(paste0(&quot;^&quot;,i,&quot;$&quot;)))&gt;0){
        BiocManager::install(i)
      }
      else{
        #  If package was not able to be loaded then re-install
        install.packages( i , dependencies = TRUE )
      }
      #  Load package after installing
      require( i , character.only = TRUE )
    }
  }
}
Pkg_list=c(&quot;devtools&quot;,
           &quot;flexdashboard&quot;,
           &quot;stringr&quot;,
           &quot;data.table&quot;,
           &quot;knitr&quot;,
           &quot;kableExtra&quot;,
           &quot;rmarkdown&quot;,
           &quot;crosstalk&quot;,
           &quot;RColorBrewer&quot;,
           &quot;gridExtra&quot;,
           &quot;grid&quot;,
           &quot;ggplot2&quot;,
           &quot;plotly&quot;,
           &quot;R.utils&quot;,
           &quot;STRINGdb&quot;,
           &quot;crosstalk&quot;,
           &quot;tidyr&quot;,
           &quot;limma&quot;,
           &quot;edgeR&quot;,
           &quot;sva&quot;,
           &quot;DT&quot;,
           &quot;pheatmap&quot;,
           &quot;org.Hs.eg.db&quot;,
           &quot;M3C&quot;,
           &quot;ggpubr&quot;,
           &quot;clusterProfiler&quot;,
           &quot;ggnewscale&quot;
)
InstallPackages(Pkg_list)</code></pre>
<pre><code>## Loading required package: devtools</code></pre>
<pre><code>## Loading required package: usethis</code></pre>
<pre><code>## Loading required package: flexdashboard</code></pre>
<pre><code>## Loading required package: rmarkdown</code></pre>
<pre><code>## Loading required package: R.utils</code></pre>
<pre><code>## Loading required package: R.oo</code></pre>
<pre><code>## Loading required package: R.methodsS3</code></pre>
<pre><code>## R.methodsS3 v1.8.1 (2020-08-26 16:20:06 UTC) successfully loaded. See ?R.methodsS3 for help.</code></pre>
<pre><code>## R.oo v1.24.0 (2020-08-26 16:11:58 UTC) successfully loaded. See ?R.oo for help.</code></pre>
<pre><code>## 
## Attaching package: &#39;R.oo&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:R.methodsS3&#39;:
## 
##     throw</code></pre>
<pre><code>## The following objects are masked from &#39;package:devtools&#39;:
## 
##     check, unload</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     trim</code></pre>
<pre><code>## The following objects are masked from &#39;package:methods&#39;:
## 
##     getClasses, getMethods</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     attach, detach, load, save</code></pre>
<pre><code>## R.utils v2.10.1 (2020-08-26 22:50:31 UTC) successfully loaded. See ?R.utils for help.</code></pre>
<pre><code>## 
## Attaching package: &#39;R.utils&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:tidyr&#39;:
## 
##     extract</code></pre>
<pre><code>## The following object is masked from &#39;package:utils&#39;:
## 
##     timestamp</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     cat, commandArgs, getOption, inherits, isOpen, nullfile, parse,
##     warnings</code></pre>
<pre><code>## Loading required package: STRINGdb</code></pre>
<pre><code>## Loading required package: DT</code></pre>
<pre><code>## Loading required package: M3C</code></pre>
<pre><code>## Loading required package: clusterProfiler</code></pre>
<pre><code>## clusterProfiler v4.0.0  For help: https://guangchuangyu.github.io/software/clusterProfiler
## 
## If you use clusterProfiler in published research, please cite:
## Guangchuang Yu, Li-Gen Wang, Yanyan Han, Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology. 2012, 16(5):284-287.</code></pre>
<pre><code>## 
## Attaching package: &#39;clusterProfiler&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:AnnotationDbi&#39;:
## 
##     select</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     slice</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     rename</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     filter</code></pre>
<pre><code>## Loading required package: ggnewscale</code></pre>
<pre class="r"><code># GitHub packages
devtools::install_github(&quot;rstudio/crosstalk&quot;)</code></pre>
<pre><code>## Skipping install of &#39;crosstalk&#39; from a github remote, the SHA1 (f8e0fa44) has not changed since last install.
##   Use `force = TRUE` to force installation</code></pre>
<pre class="r"><code>remotes::install_github(&quot;vqv/ggbiplot&quot;)</code></pre>
<pre><code>## Skipping install of &#39;ggbiplot&#39; from a github remote, the SHA1 (7325e880) has not changed since last install.
##   Use `force = TRUE` to force installation</code></pre>
<pre class="r"><code># library(data.table)
# library(knitr)
# library(kableExtra)
# # library(ggsignif)
# library(ggpubr)
# library(grid)
# library(gridExtra)
# library(RColorBrewer)
# library(org.Hs.eg.db)
# library(crosstalk)</code></pre>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
